include/SDL_video_capture.h

```diff
@@ -0,0 +1,301 @@
+
+/**
+ *  SDL Video Capture Status
+ */
+typedef struct SDL_VideoCaptureFrame
+{
+    void *pixels;               /**< Pointer to memory data */
+    int fd;                     /**< DMABUF file descriptor if available (-1 if not) */
+    void *clientbuffer;         /**< Android clientbuffer if available (NULL if not) */
+    void *internal;             /**< Private field */
+} SDL_VideoCaptureFrame;
+
+/**
+ * Open a Video Capture device
+ *
+ * \param device_name device name to open. Can be NULL.
+ *
+ * \returns device id, or SDL_VIDEO_CAPTURE_INVALID on error
+ *
+ * \sa SDL_GetVideoCaptureDeviceName
+ * \sa SDL_GetNumVideoCaptureDevices
+ */
+extern DECLSPEC SDL_VideoCaptureDeviceID SDLCALL SDL_OpenVideoCapture(const char *device_name);
+
+/**
+ * Set specification
+ *
+ * \param id video capture opened device id
+ * \param desired desired video capture spec
+ * \param obtained obtained video capture spec
+ * \param allowed_changes allow changes or not
+ *
+ * \returns 0 on success, negative value on error
+ */
+extern DECLSPEC int SDLCALL SDL_SetVideoCaptureSpec(SDL_VideoCaptureDeviceID id,
+                                                  const SDL_VideoCaptureSpec *desired,
+                                                  SDL_VideoCaptureSpec *obtained,
+                                                  int allowed_changes);
+
+/**
+ * Open a Video Capture device and set specification
+ *
+ * \param device_name device name to open. Can be NULL.
+ * \param desired desired video capture spec
+ * \param obtained obtained video capture spec
+ * \param allowed_changes allow changes or not
+ *
+ * \returns device id, or SDL_VIDEO_CAPTURE_INVALID on error
+ */
+extern DECLSPEC SDL_VideoCaptureDeviceID SDLCALL SDL_OpenVideoCaptureWithSpec(
+                                                  const char *device_name,
+                                                  const SDL_VideoCaptureSpec *desired,
+                                                  SDL_VideoCaptureSpec *obtained,
+                                                  int allowed_changes);
+/**
+ * Get device name
+ *
+ * \param index device index between 0 and num -1
+ *
+ * \returns device name, shouldn't be freed
+ *
+ * \sa SDL_GetNumVideoCaptureDevices
+ */
+extern DECLSPEC const char * SDLCALL SDL_GetVideoCaptureDeviceName(int index);
+
+/**
+ * Use this function to get the number of video capture devices.
+ *
+ * \returns the number of video capture devices
+ */
+extern DECLSPEC int SDLCALL SDL_GetNumVideoCaptureDevices(void);
+
+/**
+ * Get the obtained video capture spec
+ *
+ * \param id video capture opened device id
+ * \param spec The SDL_VideoCaptureSpec to be initialized by this function.
+ *
+ * \returns 0 on success, negative value on error
+ */
+extern DECLSPEC int SDLCALL SDL_GetVideoCaptureSpec(SDL_VideoCaptureDeviceID id, SDL_VideoCaptureSpec *spec);
+
+
+#define SDL_VIDEO_CAPTURE_TYPE_COMPRESSED   1
+#define SDL_VIDEO_CAPTURE_TYPE_EMULATED     2
```

Comment:
How does this work? Every frame format is either compressed or emulated?

I assume if a frame format is in a compressed format, an uncompressed format needs to be emulated. If you have video "compressed" in H263, you would want to emulate that as RGBA32, for instance.

Reply:
webcam can output a jpeg frame (compressed) or RGBA (emulated).
this is just some info about format that the driver give (on android I think)
maybe this could be removed if not helpfull

Reply:
I found the flags in the V4l2 docs: https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/vidioc-enum-fmt.html

I think these are separate things
You can have something compressed and not emulated
You can have something compressed and emulated

I don't think this API is planning to allow people to get compressed formats, since one specifies things in SDL_PixelFormats.

So the useful bit of information is whether the format you're looking at is "native" or "emulated," where "native" is better for performance, theoretically.

It's strange to me that V4l2 allows you to enumerate emulated video types like this. I have some experience with Microsoft Media Foundation, and I am the author of the `pygame.camera` backend for Windows. In MSMF, you do the emulation yourself using "Media Foundation Transforms." These MFTs, as they are called (no relation to NFTs), give out very little guarantees. So an MSMF function of this function would have to create an MFT of [webcam format] and then try to set the output format to everything in [formats desired] and add the ones that worked to a list. Sorry, that's kind of a ramble-- this can be ported to MSMF, it's just less convenient than the V4L2 API here.